app/ante/circuit_breaker_test.go

```diff
@@ -0,0 +1,73 @@
+package ante_test
+
+import (
+	"errors"
+	"testing"
+
+	circuitante "cosmossdk.io/x/circuit/ante"
+	upgradetypes "cosmossdk.io/x/upgrade/types"
+	cmtproto "github.com/cometbft/cometbft/proto/tendermint/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	banktypes "github.com/cosmos/cosmos-sdk/x/bank/types"
+	ibcclienttypes "github.com/cosmos/ibc-go/v8/modules/core/02-client/types"
+	"github.com/stretchr/testify/require"
+
+	"github.com/celestiaorg/celestia-app/v4/app"
+	"github.com/celestiaorg/celestia-app/v4/test/util"
+)
+
+var terminalAnteHandler = func(ctx sdk.Context, _ sdk.Tx, _ bool) (sdk.Context, error) {
+	return ctx, nil
+}
+
+func TestCircuitBreaker(t *testing.T) {
+	testApp, _ := util.SetupTestAppWithGenesisValSet(app.DefaultInitialConsensusParams())
+
+	tests := []struct {
+		name     string
+		msg      sdk.Msg
+		expError error
+	}{
+		{
+			"success: bank send",
+			&banktypes.MsgSend{},
+			nil,
+		},
+		{
+			"failure: software upgrade is blocked",
+			&upgradetypes.MsgSoftwareUpgrade{},
+			errors.New("tx type not allowed"),
+		},
+		{
+			"failure: cancel upgrade is blocked",
+			&upgradetypes.MsgCancelUpgrade{},
+			errors.New("tx type not allowed"),
+		},
+		{
+			"failure ibc software upgrade is blocked",
+			&ibcclienttypes.MsgIBCSoftwareUpgrade{},
+			errors.New("tx type not allowed"),
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			circuitAnte := circuitante.NewCircuitBreakerDecorator(&testApp.CircuitKeeper)
```

Comment:
I guess I could move this out of the loop

Reply:
https://github.com/celestiaorg/celestia-app/pull/4685/commits/a9b731f19c6fdd0ee79d8174b267d06586ab4ef9